internal/enginenetx/DESIGN.md

```diff
@@ -0,0 +1,815 @@
+
+2. a "fallback" policy.
+
+In principle, one would expect `LookupTactics` to first return all
+the tactics we can see from the stats and then try tactics obtained
+from the fallback policy. However, this simplified algorithm would
+lead to suboptimal results in the following case:
+
+1. say there are 10 tactics for "api.ooni.io:443" that are bound
+to a specific bridge address that has been discontinued;
+
+2. if we try all these 10 tactics before trying fallback tactics, we
+would waste lots of time failing before falling back.
+
+Conversely, a better strategy is to "remix" tactics as implemented
+by the [mix.go](mix.go) file:
+
+1. we take the first two tactics from the stats;
+
+2. then we take the first four tactics from the fallback;
+
+3. then we remix the rest, not caring much about whether we're
+reading from the stats of from the fallback.
+
+Because we sort tactics from the stats by our understanding of whether
+they are working as intended, we'll prioritize what we know to be working,
+but then we'll also throw some new tactics into the mix. (We read four
+tactics from the fallback because that allows us to include two bridge tactics
+and two DNS tactics, as explained below when we discuss the
+`bridgePolicy` policy.)
+
+### bridgePolicy
+
+The `bridgePolicy` is implemented by [bridgespolicy.go](bridgespolicy.go) and
+rests on the assumptions made explicit above. That is:
+
+1. that there is at least one _bridge_ for "api.ooni.io";
+
+2. that the Web Connectivity Test Helpers accepts any SNI.
+
+Here we're also using the [mix.go](mix.go) algorithm to remix
+two different sources of tactics:
+
+1. the `bridgesTacticsForDomain` only returns tactics for "api.ooni.io"
+using existing knowledge of bridges and random SNIs;
+
+2. the `maybeRewriteTestHelpersTactics` method filters the results
+coming from the fallback tactic such that, if we are connecting
+to a known test-helper domain name, we're trying to hide its SNI.
+
+The first two returned tactics will be bridges tactics for "api.ooni.io",
+if applicable, followed by two tactics generated using the DNS,
+followed by a random remix of all the remaining tactics. This choice of
+returning two and two tactics first, is the
+reason why in `statsPolicy` we return the first four tactics from
+the fallback after getting two tactics from the stats.
+
+## Overall Algorithm
+
+The composed policy is as described in Diagram 1.
+
+Therefore, the compose policy will return the following tactics:
+
+
+1. if there is a `$OONI_HOME/engine/bridges.conf` with a valid entry,
+use it without trying more tactics; otherwise,
+
+2. use the first two tactics coming from stats, if any;
+
+3. then use the first two tactics coming from bridges, if any;
+
+4. then use the first two tactics coming from the DNS, if successful;
+
+5. finally, randomly remix the remaining tactics.
+
+Excluding the case where we have a valid entry in `bridges.conf`, the following
+diagram illustrates how we're mixing tactics:
+
```

Comment:
Do I understand it correctly, that in the default thing we ship in probe we effectively have a `bridges.conf` file with the hardcoded IP?

Reply:
We have discussed this in person. Here's a recap of the conversation, for the records. The `bridges.conf` is user provided and, as such, completely overrides any other tactic. What we ship with the probe is:

1. similar to `bridges.conf` in that it hardcodes a configuration;
2. different from `bridges.conf` in that it is just one of the pieces of information used to dial.

I remember that, when we chatted, you had some suggestions regarding wording that could have helped clarify the difference. Please, provide them such that I can integrate them! 🙏

Reply:
Re-reading it I think it's fine. I'm not sure how the text can be improved if not trying to make it more concise, but I suggest we do that as future iterations.

Reply:
ack!